src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export enum PrincipalType {
+  Users = 'users',
+  Groups = 'groups',
+}
+
+export interface Principals {
+  users?: string[];
+  groups?: string[];
+}
+
+export type Permissions = Record<string, Principals>;
+
+export interface TransformedPermission {
+  type: string;
+  name: string;
+  permissions: string[];
+}
+
+const addToPrincipals = (principals?: Principals, users?: string[], groups?: string[]) => {
+  if (!principals) {
+    principals = {};
+  }
+  if (!!users) {
+    if (!principals.users) {
+      principals.users = [];
+    }
+    principals.users = Array.from(new Set([...principals.users, ...users]));
+  }
+  if (!!groups) {
+    if (!principals.groups) {
+      principals.groups = [];
+    }
+    principals.groups = Array.from(new Set([...principals.groups, ...groups]));
+  }
+  return principals;
+};
+
+const deleteFromPrincipals = (principals?: Principals, users?: string[], groups?: string[]) => {
+  if (!principals) {
+    return principals;
+  }
+  if (!!users && !!principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (!!groups && !!principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+const checkPermission = (currentPrincipals: Principals | undefined, principals: Principals) => {
```

Comment:
Just to verify: Is the user's permission always be prioritized before the group's permission? 
For example, if a user and the user's group has been given different type of permission, and the group have higher permission than the user, what will the user's permission be?

Reply:
We only check if user or the group they are associated with have the specified permission or not, so there is no difference whether the permission is from the user or associated group.